LFS-1019: Labels of answer options for vocabulary questions should be populated in the JSON

veronikaslc opened the pull request.

Commit:
LFS-1019: Labels of answer options for vocabulary questions should be populated in the JSON

marta- reviewed (commented):
Works as expected. Since it only deals with answer options for vocabulary questions, should this be reflected in the class name, e.g. `VocabularyOptionsLabelProcessor`?

Commit:
LFS-1019: Labels of answer options for vocabulary questions should be populated in the JSON
Renamed module

sashaandjic reviewed (changes requested):
tested

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/dataentry/internal/serialize/labels/VocabularyOptionsLabelProcessor.java:
```diff
@@ -0,0 +1,99 @@
+
+            for (String value : propsMap.keySet()) {
+                if (value.startsWith("/Vocabularies/")) {
+                    Node term = node.getSession().getNode(value);
```
If the vocabulary is not available, or the term has been removed in a newer version, this will fail.

```suggestion
                if (value.startsWith("/Vocabularies/") && node.getSession().nodeExists(value)) {
                    Node term = node.getSession().getNode(value);
```

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/dataentry/internal/serialize/labels/VocabularyOptionsLabelProcessor.java:
```diff
@@ -0,0 +1,99 @@
+                propsMap.put(nodeProp.getString(), nodeProp.getString());
+            }
+
+            for (String value : propsMap.keySet()) {
```
Modifying a map while iterating it is undefined behavior and may lead to a concurrent modification exception. Iterate over a clone of the key set:

```suggestion
            for (String value : new LinkedHashSet<>(propsMap.keySet())) {
```

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/dataentry/internal/serialize/labels/VocabularyOptionsLabelProcessor.java:
```diff
@@ -0,0 +1,99 @@
+                if (value.startsWith("/Vocabularies/")) {
+                    Node term = node.getSession().getNode(value);
+                    String label = term.getProperty("label").getValue().toString();
+                    if (label != null) {
```
I feel that if at this point the label is null, something else would have failed on the previous line. I would add an `if (term.hasProperty("label"))` before.

veronikaslc committed:
LFS-1019: Labels of answer options for vocabulary questions should be populated in the JSON

Fix failing if the vocabulary is not available, or the term has been removed in a newer version

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

sashaandjic commented:
![image](https://user-images.githubusercontent.com/23243244/111393618-74071a80-868f-11eb-9deb-68f4f6d088e0.png)

marta- reviewed (changes requested):
Something went wrong with the latest changes, all the labels are now replaced with the id. See excerpt from the form's  `.deep.json`:

```
				"African race": {
					"jcr:primaryType": "lfs:AnswerOption",
					"jcr:createdBy": "admin",
					"jcr:created": "2021-03-16T23:09:26.427Z",
					"value": " /Vocabularies/HANCESTRO/HANCESTRO_0010",
					"sling:resourceSuperType": "lfs/Resource",
					"jcr:uuid": "2a8fe6bb-d061-4549-a32c-d56aefc2bab0",
					"sling:resourceType": "lfs/AnswerOption",
--------------------------------------> "label": " /Vocabularies/HANCESTRO/HANCESTRO_0010"
				},
```

veronikaslc commented:
Ready for retest, rebased on latest dev.

marta- reviewed (approved):
The functionality appears to have been restored

sashaandjic commented:
retested after the merge - all good
![image](https://user-images.githubusercontent.com/23243244/111538135-b3427380-8742-11eb-8811-0633d763c101.png)
